src/rgw/rgw_rados.cc

```diff
@@ -8163,11 +8191,24 @@ int RGWRados::bucket_rebuild_index(RGWBucketInfo& bucket_info)
 {
   librados::IoCtx index_ctx;
   map<int, string> bucket_objs;
-  int r = open_bucket_index(bucket_info, index_ctx, bucket_objs);
-  if (r < 0)
+
+  /* handle on going bucket resharding */
+  int r = reshard->block_while_resharding(bucket_info.bucket.oid);
+  if (r < 0) {
     return r;
+  }
 
-  return CLSRGWIssueBucketRebuild(index_ctx, bucket_objs, cct->_conf->rgw_bucket_index_max_aio)();
+  r = open_bucket_index(bucket_info, index_ctx, bucket_objs);
+  if (r < 0) {
+    reshard->unlock_bucket_index(bucket_info.bucket.oid);
```

Comment:
@oritwas there are a bunch of calls to unlock_bucket_index() that were added and are not conditional on the resharding status. Do we now lock the bucket index anyway? I don't think we should. Anyway, it'd be less error prone to have unlock_bucket_index() triggered automatically when existing the methods. Will need to do similar to what we do with `Mutex::Locker` or with other similar cases.